File: 3/marek_kwasny_tests/marek_kwasny_tests.pl
% Definiujemy moduł zawierający testy.
% Należy zmienić nazwę modułu na {imie}_{nazwisko}_tests gdzie za
% {imie} i {nazwisko} należy podstawić odpowiednio swoje imię
% i nazwisko bez wielkich liter oraz znaków diakrytycznych
:- module(marek_kwasny_tests, [tests/3]).

% Zbiór faktów definiujących testy
% Należy zdefiniować swoje testy
tests(empty_program, input(""), program([])).
tests(invalid, input("def main()"), no).
tests(adder, file('adder.hdml'), yes).
tests(srcpos, input("def main(_) = 1"),
  program([def(main, wildcard(file(test, 1, 10, 9, 1)), num(no, 1))])).

tests(empty_from_file, file('empty.hdml'), yes).
tests(bad_function_name, file('bad_function_name.hdml'), no).
tests(hdml_feature, file('hdml_feature.hdml'), no).
tests(bad_if, file('bad_if.hdml'), no).
tests(example, input("def half_adder(A, B) = A & B, A ^ B"), 
	program([def(half_adder, pair(no, var(no, 'A'), var(no, 'B')), pair(no, op(no, '&', var(no, 'A'), var(no, 'B')), op(no, '^', var(no, 'A'), var(no, 'B'))))])).

tests(test1, file('test1.hdml'), yes).
tests(test2, file('test2.hdml'), yes).
tests(test3, file('test3.hdml'), yes).

tests(test4, input("def f(X) = X + 2"),
	program([def(f, var(no, 'X'), op(no, '+', var(no, 'X'), num(no, 2)))])). 

tests(test5, input("def f(_) = []"), 
	program([def(f, wildcard(no), empty(no))])).

tests(test6, input("def f(X, Y) = let _ = f(0, 0) in []"),
	program([def(f, pair(no, var(no, 'X'), var(no, 'Y')), let(no, wildcard(no), call(no, 'f', pair(no, num(no, 0), num(no, 0))), empty(no)))])).

tests(test7, file('test7.hdml'), no).
tests(test8, file('test8.hdml'), yes).
tests(test9, input("def f((((_)))) = []"), yes).
tests(test10, file('test10.hdml'), yes).

File: 4/marek_kwasny_tests/MarekKwasnyTests.hs
-- Wymagamy, by moduł zawierał tylko bezpieczne funkcje
{-# LANGUAGE Safe #-}
-- Definiujemy moduł zawierający testy.
-- Należy zmienić nazwę modułu na {Imie}{Nazwisko}Tests gdzie za {Imie}
-- i {Nazwisko} należy podstawić odpowiednio swoje imię i nazwisko
-- zaczynające się wielką literą oraz bez znaków diakrytycznych.
module MarekKwasnyTests(tests) where

-- Importujemy moduł zawierający typy danych potrzebne w zadaniu
import DataTypes

-- Lista testów do zadania
-- Należy uzupełnić jej definicję swoimi testami
tests :: [Test]
tests =
  [ Test "inc"          (SrcString "input x in x + 1")                                         (Eval [42] (Value 43))
  , Test "undefVar"     (SrcString "x")                                                        TypeError
  , Test "adding"       (SrcString "1 + 3 + 2")                                                (Eval [] (Value 6))
  , Test "ifneg"        (SrcString "input x y in if x > y then 1 else 0")                      (Eval [1, 2] (Value 0))
  , Test "iftrue"       (SrcString "input x y in if x > y then 1 else 0")                      (Eval [2, 1] (Value 1))
  , Test "let"          (SrcString "input y in let x = 10 in if y < x then 1 else 0")          (Eval [5] (Value 1))
  , Test "boolint"      (SrcString "let x = true in if x > 2 then 1 else 0")                   TypeError
  , Test "nonint"       (SrcString "2 + true")                                                 TypeError
  , Test "nonbool"      (SrcString "input x in if 5 then false else x")                        TypeError
  , Test "undef"        (SrcString "input x in x + y")                                         TypeError
  , Test "ifinifelse"   (SrcString "input x y in if x > 2 then if y > 2 then 1 else 0 else 2") (Eval [3, 1] (Value 0))
  , Test "ifinifskip"   (SrcString "input x y in if x > 2 then if y > 2 then 1 else 0 else 2") (Eval [1, 0] (Value 2))
  , Test "ifinifthen"   (SrcString "input x y in if x > 2 then if y > 2 then 1 else 0 else 2") (Eval [3, 3] (Value 1))
  , Test "ifinelseif1"  (SrcString "input x y in if x > 2 then 0 else if y > 2 then 1 else 2") (Eval [3, 5] (Value 0))
  , Test "ifinelseif2"  (SrcString "input x y in if x > 2 then 0 else if y > 2 then 1 else 2") (Eval [1, 5] (Value 1))
  , Test "ifinelseif3"  (SrcString "input x y in if x > 2 then 0 else if y > 2 then 1 else 2") (Eval [1, 1] (Value 2))
  , Test "modinlet"     (SrcString "input y in let x = y mod 2 in if x > 0 then 0 else 1")     (Eval [4] (Value 1))
  ]